### PR TITLE
mariadb: 10.3.15 -> 10.3.17

### PR DESCRIPTION
**Pull request by Izorkin**
###### Motivation for this change
Update MariaDB to 10.3.16
Changelog - https://mariadb.com/kb/en/library/mariadb-10316-changelog/

Move mariadb_config and mysql_config to $dev output.
Move mysql binary to mysql.client

Fixed mysql_config output
before PR
```
Usage: ./result/bin/mysql_config [OPTIONS]
Options:
        --cflags         [-I/nix/store/63l1lwz19lj4mw03m2ra7m1vnylp8xxl-mariadb-10.3.15/include/mysql -I/nix/store/63l1lwz19lj4mw03m2ra7m1vnylp8xxl-mariadb-10.3.15/include/mysql/mysql]
        --include        [-I/nix/store/63l1lwz19lj4mw03m2ra7m1vnylp8xxl-mariadb-10.3.15/include/mysql -I/nix/store/63l1lwz19lj4mw03m2ra7m1vnylp8xxl-mariadb-10.3.15/include/mysql/mysql]
        --libs           [-L/nix/store/63l1lwz19lj4mw03m2ra7m1vnylp8xxl-mariadb-10.3.15/lib/ -lmariadb -lz -ldl -lm -lpthread -lssl -lcrypto]
        --libs_r         [-L/nix/store/63l1lwz19lj4mw03m2ra7m1vnylp8xxl-mariadb-10.3.15/lib/ -lmariadb -lz -ldl -lm -lpthread -lssl -lcrypto]
        --plugindir      [/nix/store/63l1lwz19lj4mw03m2ra7m1vnylp8xxl-mariadb-10.3.15/lib/mysql/plugin]
        --socket         [/run/mysqld/mysqld.sock]
        --port           [3306]
        --version        [10.3.15]
        --libmysqld-libs [-L/home/home/works/src-nix/nixpkgs/result/lib/mysql  -lmysqld]
        --variable=VAR   VAR is one of:
                pkgincludedir [/nix/store/63l1lwz19lj4mw03m2ra7m1vnylp8xxl-mariadb-10.3.15/include/mysql]
                pkglibdir     [/home/home/works/src-nix/nixpkgs/result/lib/mysql]
                plugindir     [/nix/store/63l1lwz19lj4mw03m2ra7m1vnylp8xxl-mariadb-10.3.15/lib/mysql/plugin]
```
After:
```
Usage: ./result-dev/bin/mysql_config [OPTIONS]
Options:
        --cflags         [-I/nix/store/w9569b5rpvnma3iz55il9y6hwbpi2sgq-mariadb-client-10.3.16-dev/include/mysql -I/nix/store/w9569b5rpvnma3iz55il9y6hwbpi2sgq-mariadb-client-10.3.16-dev/include/mysql/mysql]
        --include        [-I/nix/store/w9569b5rpvnma3iz55il9y6hwbpi2sgq-mariadb-client-10.3.16-dev/include/mysql -I/nix/store/w9569b5rpvnma3iz55il9y6hwbpi2sgq-mariadb-client-10.3.16-dev/include/mysql/mysql]
        --libs           [-L/nix/store/w9569b5rpvnma3iz55il9y6hwbpi2sgq-mariadb-client-10.3.16-dev/lib/mysql/ -lmariadb -lz -ldl -lm -lpthread -lssl -lcrypto]
        --libs_r         [-L/nix/store/w9569b5rpvnma3iz55il9y6hwbpi2sgq-mariadb-client-10.3.16-dev/lib/mysql/ -lmariadb -lz -ldl -lm -lpthread -lssl -lcrypto]
        --plugindir      [/nix/store/w9569b5rpvnma3iz55il9y6hwbpi2sgq-mariadb-client-10.3.16-dev/lib/mysql/plugin]
        --socket         [/run/mysqld/mysqld.sock]
        --port           [3306]
        --version        [10.3.16]
        --libmysqld-libs [-L/home/home/works/src-nix/nixpkgs/result-dev/lib/mysql  -lmysqld]
        --variable=VAR   VAR is one of:
                pkgincludedir [/home/home/works/src-nix/nixpkgs/result-dev/include/mysql]
                pkglibdir     [/home/home/works/src-nix/nixpkgs/result-dev/lib/mysql]
                plugindir     [/nix/store/w9569b5rpvnma3iz55il9y6hwbpi2sgq-mariadb-client-10.3.16-dev/lib/mysql/plugin]
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb.client mariadb.server
@GrahamcOfBorg test mysql mysqlBackup

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb.client mariadb.server
@GrahamcOfBorg test mysql mysqlBackup

**Comment by matthewbauer** [comment]
Looks okay, but I want to make sure x86_64-darwin builds.

@GrahamcOfBorg build mariadb.client mariadb.server

File: pkgs/servers/sql/mariadb/cmake-libmariadb-includedir.patch [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+--- a/libmariadb/mariadb_config/libmariadb.pc.in
++++ b/libmariadb/mariadb_config/libmariadb.pc.in
+@@ -6,7 +6,7 @@
+ #  Dan Nicholson’s Guide to pkg-config (http://www.freedesktop.org/wiki/Software/pkg-config/)
+ #
+ 
+-prefix=@CMAKE_INSTALL_PREFIX@
++prefix=@CMAKE_INSTALL_PREFIX_DEV@
+ includedir=${prefix}/@INSTALL_INCLUDEDIR@/@SUFFIX_INSTALL_DIR@
+ libdir=${prefix}/@INSTALL_LIBDIR@/@SUFFIX_INSTALL_DIR@
```
Comment (matthewbauer) [inline pkgs/servers/sql/mariadb/cmake-libmariadb-includedir.patch]: Shouldn't this one be in "out"?

**Comment by Izorkin** [comment]
@matthewbauer need upgrade MariaDB to version 10.4.6 - current stable release?

File: pkgs/servers/sql/mariadb/cmake-libmariadb-includedir.patch (hunk 1 above)
Comment (Izorkin) [reply pkgs/servers/sql/mariadb/cmake-libmariadb-includedir.patch]: These files are used to build other packages. Checked by file structure in mariad connector-c.
Packages mariadb-server-dev = mariadb-client-dev = mariadb-connector-c

**Comment by Izorkin** [comment]
Do you need these binaries:
- mysql_client_test
- mytop (not worked)
- mysqltest

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb.client mariadb.server
@GrahamcOfBorg test mysql mysqlBackup

**Comment by Izorkin** [comment]
added release notes

**Comment by Izorkin** [comment]
Update configuration. Please recheck.

**Comment by Mic92** [comment]
@GrahamcOfBorg build mariadb

File: pkgs/servers/sql/mariadb/default.nix [hunk 2]
```diff
@@ -27,30 +28,32 @@ galeraLibs = buildEnv {
...
 
-  patches = [ ./cmake-includedir.patch ]
-    ++ optionals stdenv.isDarwin [
+  patches = [
+    ./cmake-includedir.patch
+    ./cmake-libmariadb-includedir.patch
+  ] ++ optionals stdenv.isDarwin [
       # Derived from "Fixed c++11 narrowing error"
       # https://github.com/MariaDB/server/commit/a0dfefb0f8a47145e599a5f1b0dc576fa7634b92
       ./fix-c++11-narrowing-error.patch
```
Comment (Mic92) [inline pkgs/servers/sql/mariadb/default.nix]: This patch is now upstream.

Comment (Izorkin) [reply pkgs/servers/sql/mariadb/default.nix]: fixed

**Comment by Mic92** [comment]
@GrahamcOfBorg build mariadb

**Comment by Izorkin** [comment]
@Mic92 need add mariadb 10.4.7 separate package?

**Comment by Mic92** [comment]
@Izorkin if possible I would like to avoid that, but I don't know the mariadb versioning well to reason about potential breakages between these two versions.

**Comment by Mic92** [comment]
Looking at the current build error on macOS it could be that openssl has to be appear earlier in the linker flags as the linker does not pick-up all symbols from it. cc @matthewbauer @LnL7

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb.client mariadb.server

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb.connector-c_3_1

**Comment by Mic92** [comment]
@GrahamcOfBorg build mariadb.connector-c_3_1

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb.connector-c_3_1

**Comment by Mic92** [comment]
@GrahamcOfBorg build mariadb.connector-c_3_1

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb.connector-c_3_1 mariadb.client mariadb.server

**Comment by Mic92** [comment]
@GrahamcOfBorg build mariadb.connector-c_3_1 mariadb.client mariadb.server

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb.connector-c_3_1

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb.client

**Comment by Mic92** [comment]
@GrahamcOfBorg build mariadb.connector-c_3_1

**Comment by Mic92** [comment]
@GrahamcOfBorg build mariadb.client

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb.client mariadb.server

**Comment by Mic92** [comment]
@GrahamcOfBorg build mariadb.client mariadb.server

**Comment by Izorkin** [comment]
@GrahamcOfBorg build mariadb.client mariadb.server

**Comment by Mic92** [comment]
@GrahamcOfBorg build mariadb.client mariadb.server

**Comment by Izorkin** [comment]
Thanks!

**Comment by orivej** [comment]
Why did you move `libmysqlclient.so` from `lib` to `lib/mysql`? It seems wrong.

File: pkgs/servers/sql/mariadb/cmake-libmariadb-includedir.patch [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+--- a/libmariadb/mariadb_config/mariadb_config.c.in
++++ b/libmariadb/mariadb_config/mariadb_config.c.in
+@@ -5,13 +5,13 @@
+ 
+ static char *mariadb_progname;
+ 
+-#define INCLUDE "-I@CMAKE_INSTALL_PREFIX@/@INSTALL_INCLUDEDIR@ -I@CMAKE_INSTALL_PREFIX@/@INSTALL_INCLUDEDIR@/mysql"
+-#define LIBS    "-L@CMAKE_INSTALL_PREFIX@/@INSTALL_LIBDIR@/ -lmariadb"
++#define INCLUDE "-I@CMAKE_INSTALL_PREFIX_DEV@/@INSTALL_INCLUDEDIR@ -I@CMAKE_INSTALL_PREFIX_DEV@/@INSTALL_INCLUDEDIR@/mysql"
++#define LIBS    "-L@CMAKE_INSTALL_PREFIX_DEV@/@INSTALL_LIBDIR@/ -lmariadb"
```
Comment (orivej) [inline pkgs/servers/sql/mariadb/cmake-libmariadb-includedir.patch]: This makes `mariadb_config --libs` list the directory with static libraries.

**Comment by Izorkin** [comment]
> Why did you move `libmysqlclient.so` from `lib` to `lib/mysql`? It seems wrong.

In package mariadb.connector-c file libmysqlclient.so located in lib/mysql/
```
ls -lah ./result/lib/mysql/ | grep libmysqlclient.so 
lrwxrwxrwx 31 root root   13 янв  1  1970 libmysqlclient.so -> libmariadb.so
```
Made similarly here.
Or is it wrong?

File: pkgs/servers/sql/mariadb/cmake-libmariadb-includedir.patch (hunk 3 above)
Comment (Izorkin) [reply pkgs/servers/sql/mariadb/cmake-libmariadb-includedir.patch]: But how should it be?
Now mariadb.connector-c is used instead of mariadb.client.dev

**Comment by orivej** [comment]
I guess that connector-c default libdir is `lib/mariadb` while client default is `lib` to prevent file conflicts between their packages in traditional distros. Libraries in `lib` are more useful because they can be used without configuration.

**Comment by Izorkin** [comment]
@orivej  use this patch?
```
diff --git a/pkgs/servers/sql/mariadb/default.nix b/pkgs/servers/sql/mariadb/default.nix
index 7d34dd8d63b..e92f3585624 100644
--- a/pkgs/servers/sql/mariadb/default.nix
+++ b/pkgs/servers/sql/mariadb/default.nix
@@ -68,8 +68,6 @@ common = rec { # attributes common to both builds
     "-DINSTALL_DOCDIR=share/doc/mysql"
     "-DINSTALL_DOCREADMEDIR=share/doc/mysql"
     "-DINSTALL_INCLUDEDIR=include/mysql"
-    "-DINSTALL_LIBDIR=lib/mysql"
-    "-DINSTALL_PLUGINDIR=lib/mysql/plugin"
     "-DINSTALL_INFODIR=share/mysql/docs"
     "-DINSTALL_MANDIR=share/man"
     "-DINSTALL_MYSQLSHAREDIR=share/mysql"
@@ -95,10 +93,10 @@ common = rec { # attributes common to both builds
   ];

   postInstall = ''
-    rm "$out"/lib/mysql/plugin/daemon_example.ini
+    rm "$out"/lib/plugin/daemon_example.ini
     mkdir -p "$dev"/bin && mv "$out"/bin/{mariadb_config,mysql_config} "$dev"/bin
-    mkdir -p "$dev"/lib/mysql && mv "$out"/lib/mysql/{libmariadbclient.a,libmysqlclient.a,libmysqlclient_r.a,libmysqlservices.a} "$dev"/lib/mysql
-    mkdir -p "$dev"/lib/mysql/plugin && mv "$out"/lib/mysql/plugin/{caching_sha2_password.so,dialog.so,mysql_clear_password.so,sha256_password.so} "$dev"/lib/mysql/plugin
+    mkdir -p "$dev"/lib/ && mv "$out"/lib/{libmariadbclient.a,libmysqlclient.a,libmysqlclient_r.a,libmysqlservices.a} "$dev"/lib
+    mkdir -p "$dev"/lib/plugin && mv "$out"/lib/plugin/{caching_sha2_password.so,dialog.so,mysql_clear_password.so,sha256_password.so} "$dev"/lib/plugin
   '';

   enableParallelBuilding = true;
@@ -136,14 +134,13 @@ client = stdenv.mkDerivation (common // {
       -DCMAKE_INSTALL_PREFIX_DEV=$dev"
   '';

-  postInstall =  common.postInstall + ''
+  postInstall = common.postInstall + ''
     rm -r "$out"/share/doc
     rm "$out"/bin/{mysqltest,mytop,wsrep_sst_rsync_wan}
-    libmysqlclient_path=$(readlink -f $out/lib/mysql/libmysqlclient${libExt})
-    rm "$out"/lib/mysql/{libmariadb${libExt},libmysqlclient${libExt},libmysqlclient_r${libExt}}
-    mv "$libmysqlclient_path" "$out"/lib/mysql/libmysqlclient${libExt}
-    ln -sv libmysqlclient${libExt} "$out"/lib/mysql/libmysqlclient_r${libExt}
-
+    libmysqlclient_path=$(readlink -f $out/lib/libmysqlclient${libExt})
+    rm "$out"/lib/{libmariadb${libExt},libmysqlclient${libExt},libmysqlclient_r${libExt}}
+    mv "$libmysqlclient_path" "$out"/lib/libmysqlclient${libExt}
+    ln -sv libmysqlclient${libExt} "$out"/lib/libmysqlclient_r${libExt}
   '';
 });

@@ -166,7 +163,6 @@ server = stdenv.mkDerivation (common // {

   cmakeFlags = common.cmakeFlags ++ [
     "-DMYSQL_DATADIR=/var/lib/mysql"
-    "-DINSTALL_PLUGINDIR=lib/mysql/plugin"
     "-DENABLED_LOCAL_INFILE=OFF"
     "-DWITH_READLINE=ON"
     "-DWITH_EXTRA_CHARSETS=all"
@@ -196,15 +192,15 @@ server = stdenv.mkDerivation (common // {
     chmod +x "$out"/bin/wsrep_sst_common
     rm "$out"/bin/mysql_client_test
     rm -r "$out"/data # Don't need testing data
-    rm "$out"/lib/mysql/{libmysqlclient${libExt},libmysqlclient_r${libExt}}
+    rm "$out"/lib/{libmysqlclient${libExt},libmysqlclient_r${libExt}}
     mv "$out"/share/{groonga,groonga-normalizer-mysql} "$out"/share/doc/mysql
   '' + optionalString withoutClient ''
     ${ # We don't build with GSSAPI on Darwin
       optionalString (!stdenv.isDarwin) ''
-        rm "$out"/lib/mysql/plugin/auth_gssapi_client.so
+        rm "$out"/lib/plugin/auth_gssapi_client.so
       ''
     }
-    rm "$out"/lib/mysql/plugin/client_ed25519.so
+    rm "$out"/lib/plugin/client_ed25519.so
   '' + optionalString (!stdenv.isDarwin) ''
     sed -i 's/-mariadb/-mysql/' "$out"/bin/galera_new_cluster
   '';
```

**Comment by orivej** [comment]
mariadb (and mysql) defines a set of install layouts: standalone, rpm, deb, svr4: https://github.com/MariaDB/server/blob/ae4b9b76/cmake/install_layout.cmake#L76 . We configure it as `standalone`, which is meant for read-writable single directory installations, but then override standalone defaults towards `deb`, which is closer to what we need in nixpkgs. deb install libs in `lib` and plugins in `lib/mysql/plugin`: https://github.com/MariaDB/server/blob/ae4b9b76/cmake/install_layout.cmake#L177-L178 .

I suppose that it would be easier for us to switch the base layout to `deb`, but meanwhile its fine to tweak `standalone` to be more like `deb` and use `lib` and `lib/mysql/plugin`.

**Comment by Izorkin** [comment]
@orivej updated patch

**Comment by Izorkin** [comment]
@orivej create new PR with this patch?

**Pull request merged**